### DISCUSSION TITLE
Multiple dark/light theme

### DISCUSSION THREAD

**Discussion by anonymouscatcher** [chakra-ui/chakra-ui#3873] [Q&A]
Hello Everyone, 

I'm trying to build an application, It supports darks/lights themes (like Slacks theme), the base is dark or light but some backgrounds or colours are different.
I migrate to chakra v1  recentlyI found it's kind of hard to do since it has different API for managing the theme. So my first issue is when you define baseStyle for a component you can only access to colorMode which is good but not enough, I'm looking for a way to pass a "sub theme name", Imagine we have these themes, 

![image](https://user-images.githubusercontent.com/53314423/115520689-b58c7600-a28a-11eb-9a0c-1de812639121.png)

I need to define a heading for each,
```js
import { extendTheme } from '@chakra-ui/react';

const config = {
  initialColorMode: 'dark',
  useSystemColorMode: false,
};

const getBgBasedOnTheme = ({subThemeNameFromArgs}) => {
   switch (subThemeNameFromArgs) {
     case 'subDarkTheme1':
     case: 'subDarkTheme2'
     case: 'subDarkTheme3'
   }
};

const components = {
  HeaderApp: {
    baseStyle: ({ colorMode, ...rest }) => {
      console.log(colorMode, rest); // returns dark or light
      // rest has also useless properties
      return {
         bg: getBgBasedOnTheme(rest)
      };
    }
  }
};

export const theme = extendTheme({
  config,
  components
});
```

I can handle this with store management tricks and take these logic to a context or something but I'm sure there should be a cleaner and performant way.

In the past I solved the issue with having different theme objects like this:
```js
export const theme = {
  dark,
  light
};


const GlobalProviders = ({ children }) => {
 // some states to change the theme
  return (
    <ChakraProvider theme={theme.dark}>
      {children}
    </ChakraProvider>
  );
};
```
But the issue was you needed to write css for each component. I like the new approach that chakra has now.
Thanks in advance

**Comment by anonymouscatcher** [comment]
I could manage this by passing props to theme it self then you can detect it from the `rest.theme`.

```js
export const theme = extendTheme({
  config,
  subTheme: 'subDarkTheme1', // of course you need to pass it to theme provider and change it dynamically 
  components
});
```

But still have issue because `baseStyle()` doesn't get call when you pass it as a function:
```js
const IconButton = chakra(ChakraButton, {
  baseStyle({theme}) {
    console.log(theme);
    return {
      fontWeight: 'bold',
      textTransform: 'uppercase',
      borderRadius: 'base',
      bg: 'red'
    };
  }
});
```

**Comment by TimKolberger** [comment]
Do you only want to override the colors depending on the users choice?
If so I would extend your base theme with your desired overrides for each color palette and switch the theme object for the ChakraProvider.

```ts
import { extendTheme } from '@chakra-ui/react'

const yourBaseTheme = extendTheme(yourCommonOverrides)

const eggplantTheme =  extendTheme({ colors: eggplantThemeColors }, yourBaseTheme)
const cmykTheme =  extendTheme({ colors: cmykThemeColors }, yourBaseTheme)
// ...

export const App = () => {
  const currentTheme = eggplantTheme /* add logic to change the theme here */
  return <ChakraProvider theme={currentTheme}>{/*...*/}</ChakraProvider>
}
```

PS: I am closing the referenced issue, because this is a question - not an issue :)

**Reply by EliasLeguizamon123** [reply]
Hi ! @TimKolberger i have same question here, i am new using chakra (and i love it by the way), so i have this theme.js file :

```
import { extendTheme } from '@chakra-ui/react';

const theme = extendTheme({
    colors: {
        primary: '#7FC2C2',
        secondary: '#E1774B',
        tertiary: '#F4DB60',
    },
    config: {
        initialColorMode: 'light',
    },
});

const samurai = extendTheme({
    colors: {
        primary: '#000000',
        secondary: '#989C94',
        tertiary: '#C21515',
    },
    config: {
        initialColorMode: 'dark',
    },
});

export default { theme, samurai };

```

also in my index.js i have this code :

```
import ReactDOM from 'react-dom';
import './index.css';
import App from './App';
import reportWebVitals from './reportWebVitals';
import { ChakraProvider } from '@chakra-ui/react';
import { theme, samurai } from './theme';

const currentTheme = samurai;

ReactDOM.render(
    <ChakraProvider theme={currentTheme}>
        <App />
    </ChakraProvider>,
    document.getElementById('root')
);

```
But it doesnt work how i expect, you have any suggestion to change it correctly ?

**Reply by patmifsud** [reply]
Hi Elias, we’re you able to get this working in the end? We have the exact same usecase and can’t get the themes to swap without refreshing the page

**Reply by EliasLeguizamon123** [reply]
Hi @patmifsud no, i cant make it works.

**Reply by TimKolberger** [reply]
I see that my last example is missing the juicy part.

`/* add logic to change the theme here */` is a placeholder for a component state to manage which theme is currently selected.

A minimal working example is `React.useState`.

**The limitation is that the selection won't survive a page reload.**

```tsx
 export default function App() {
  const themes = {
    red: redTheme,
    green: greenTheme
  };

  // you could persist this to localstorage with e.g. https://usehooks.com/useLocalStorage/
  // if the selection should survive a page reload
  const [currentThemeName, setCurrentThemeName] = React.useState<
    "red" | "green"
  >("red");

  function toggleTheme() {
    setCurrentThemeName((prevThemeName) =>
      prevThemeName === "red" ? "green" : "red"
    );
  }

  const currentTheme = themes[currentThemeName];

  return (
    <ChakraProvider theme={currentTheme}>
      <div>
        <Box color="brand.500">I am in brand.500 color</Box>
        <Button onClick={toggleTheme}>Toggle Theme</Button>
      </div>
    </ChakraProvider>
  );
}
```

[See this CodeSandbox for a working example](https://codesandbox.io/s/chakra-ui-multiple-theme-toggle-ofn6zm?file=/src/App.tsx:411-604)
